tests/unit/test_upgrade.py

```diff
@@ -0,0 +1,258 @@
+        harness.update_relation_data(
+            harness.charm.state.peer_relation.id,
+            f"{CHARM_KEY}/2",
+            {"private-address": "222.222.222"},
+        )
+
+    stack = harness.charm.upgrade.build_upgrade_stack()
+
+    assert len(stack) == 0
+    # assert len(stack) == len(set(stack))
+
+
+@pytest.mark.parametrize("upgrade_stack", ([], [0]))
+def test_run_password_rotation_while_upgrading(harness, upgrade_stack):
+    harness.charm.upgrade.upgrade_stack = upgrade_stack
+    harness.set_leader(True)
+
+    mock_event = MagicMock()
+    mock_event.params = {"username": "admin"}
+
+    with (
+        patch("charm.KafkaCharm.healthy", new_callable=PropertyMock, return_value=True),
+        patch("managers.auth.AuthManager.add_user"),
+    ):
+        harness.charm.password_action_events._set_password_action(mock_event)
+
+    if not upgrade_stack:
+        mock_event.set_results.assert_called()
+    else:
+        mock_event.fail.assert_called_with(
+            f"Cannot set password while upgrading (upgrade_stack: {upgrade_stack})"
+        )
+
+
+def test_kafka_dependency_model():
+    assert sorted(KafkaDependencyModel.__fields__.keys()) == sorted(DEPENDENCIES.keys())
+
+    for value in DEPENDENCIES.values():
+        assert DependencyModel(**value)
+
+
+def test_upgrade_granted_sets_failed_if_zookeeper_dependency_check_fails(harness: Harness):
+    with (
+        patch.object(KazooClient, "start"),
+        patch(
+            "core.models.ZooKeeperManager.get_leader",
+            new_callable=PropertyMock,
+            return_value="000.000.000",
+        ),
+        # NOTE: Dependency requires >3
+        patch(
+            "core.models.ZooKeeper.zookeeper_version",
+            new_callable=PropertyMock,
+            return_value="1.2.3",
+        ),
+        patch(
+            "core.cluster.ClusterState.ready_to_start",
+            new_callable=PropertyMock,
+            return_value=True,
+        ),
+        patch(
+            "events.upgrade.KafkaUpgradeEvents.idle",
+            new_callable=PropertyMock,
+            return_value=False,
+        ),
+    ):
+        mock_event = MagicMock()
+        harness.charm.upgrade._on_kafka_pebble_ready_upgrade(mock_event)
+
+    assert harness.charm.upgrade.state == "failed"
+
+
+# def test_upgrade_granted_sets_failed_if_failed_snap(harness: Harness):
+#     with (
+#         patch(
+#             "events.upgrade.KafkaUpgrade.zookeeper_current_version",
+#             new_callable=PropertyMock,
+#             return_value="3.6",
+#         ),
+#         patch("workload.KafkaWorkload.stop") as patched_stop,
+#         patch("workload.KafkaWorkload.install", return_value=False),
+#     ):
+#         mock_event = MagicMock()
+#         harness.charm.upgrade._on_upgrade_granted(mock_event)
+#
+#     patched_stop.assert_called_once()
+#     assert harness.charm.upgrade.state == "failed"
```

Comment:
**question:** Why are these tests commented out? Is it because they don't apply to K8s? If so:

**todo:** Make this file the same as on VM, and set `pytest.mark.skipif(substrate=foo)`

Reply:
We were skipping upgrade unittests all together but I wanted to have the unittest for the ticket in scope of the PR. While doing this I tried to copy the tests from vm and adapt for what it was sensible to. Fine to use the decorator instead of commenting